feat(ci): add `fmt` workflow

MaxMustermann2 opened the pull request:
- Add a new `fmt` workflow that runs `forge fmt --check`
- Run it in the repo root and commit
- Respond to code rabbit's review comments to remove duplicate code for vault addition or removal between `ClientChainGateway` and `Bootstrap`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Introduced GitHub Actions workflows for building, testing, and formatting code using the Foundry toolchain.
  - Updated `.gitignore` to exclude the `.gas-snapshot` directory.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bwhour commented on .prettierrc:
```diff
@@ -0,0 +1,17 @@
+  "overrides": [
+    {
+      "files": "*.sol",
+      "options": {
```
we can add "parser": "solidity-parse" to better support solidity files.

MaxMustermann2 commented on src/core/CustomProxyAdmin.sol:
```diff
@@ -16,19 +16,10 @@ contract CustomProxyAdmin is Initializable, ProxyAdmin {
-            msg.sender == proxy,
-            "CustomProxyAdmin: sender must be the proxy itself"
-        );
+    function changeImplementation(address proxy, address implementation, bytes memory data) public virtual {
+        require(msg.sender == bootstrapper, "CustomProxyAdmin: sender must be bootstrapper");
+        require(msg.sender == proxy, "CustomProxyAdmin: sender must be the proxy itself");
```
This interpretation is inaccurate.

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -299,7 +244,7 @@ contract Bootstrap is
      * public key is not found among the registered operators, indicating that the key
      * is unique and can be safely used for a new or updating operator.
-    */
+     */
```
This line already exists. Skip.

MaxMustermann2 commented on src/storage/ClientChainGatewayStorage.sol:
```diff
@@ -66,19 +66,28 @@ contract ClientChainGatewayStorage is BootstrapStorage {
 
     modifier isValidBech32Address(string calldata exocoreAddress) {
-        require(isValidExocoreAddress(exocoreAddress), "BaseRestakingController: invalid bech32 encoded Exocore address");
+        require(
+            isValidExocoreAddress(exocoreAddress),
+            "BaseRestakingController: invalid bech32 encoded Exocore address"
+        );
```
It is not reasonable to implement bech32 in Solidity. Skip.

adu-web3 commented:
if I format the files with `forge fmt`, would it pass this check?

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -346,44 +291,30 @@ contract Bootstrap is
-    */
+     */
     function nameInUse(string memory newName) public view returns (bool) {
         for (uint256 i = 0; i < registeredOperators.length; i++) {
             address ethAddress = registeredOperators[i];
             string memory exoAddress = ethToExocoreAddress[ethAddress];
-            if (keccak256(abi.encodePacked(operators[exoAddress].name)) ==
-                keccak256(abi.encodePacked(newName))) {
+            if (keccak256(abi.encodePacked(operators[exoAddress].name)) == keccak256(abi.encodePacked(newName))) {
```
This is easily possible to implement but will produce another mapping to clear later (or leave behind, considering the gas costs of clearing). Not worth it IMO but I will leave this open in case anyone feels otherwise.

MaxMustermann2 replied:
Can you make an issue about this?

MaxMustermann2 commented:
> if I format the files with `forge fmt`, would it pass this check?

No, `prettier` and `forge fmt` conflict. I have removed `prettier` and moved everything to `forge fmt`.

MaxMustermann2 commented on test/foundry/Bootstrap.t.sol:
```diff
@@ -158,10 +160,7 @@ contract BootstrapTest is Test {
-            address(clientGatewayLogic),
-            initialization
-        );
+        bootstrap.setClientChainGatewayLogic(address(clientGatewayLogic), initialization);
```
This has been actioned by requiring that the address supplied is non-zero and that the length of the initialization data is >= 4 (for the function selector).

MaxMustermann2 commented:
@coderabbitai review

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -346,44 +261,30 @@ contract Bootstrap is
-    */
+     */
     function nameInUse(string memory newName) public view returns (bool) {
         for (uint256 i = 0; i < registeredOperators.length; i++) {
             address ethAddress = registeredOperators[i];
             string memory exoAddress = ethToExocoreAddress[ethAddress];
-            if (keccak256(abi.encodePacked(operators[exoAddress].name)) ==
-                keccak256(abi.encodePacked(newName))) {
+            if (keccak256(abi.encodePacked(operators[exoAddress].name)) == keccak256(abi.encodePacked(newName))) {
```
Repeat of https://github.com/ExocoreNetwork/exocore-contracts/pull/16#discussion_r1624000782

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -237,49 +180,25 @@
-            isValidExocoreAddress(operatorExocoreAddress),
-            "Bootstrap: invalid bech32 address"
-        );
+        require(isValidExocoreAddress(operatorExocoreAddress), "Bootstrap: invalid bech32 address");
```
The `operatorExocoreAddress` check comment is a repeat of https://github.com/ExocoreNetwork/exocore-contracts/pull/16#discussion_r1624000788. Others are already implemented.

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -299,7 +218,7 @@
      * public key is not found among the registered operators, indicating that the key
      * is unique and can be safely used for a new or updating operator.
-    */
+     */
```
The function is kept `public` intentionally so that others can call it to validate their inputs.

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -593,25 +476,23 @@
+        public
+        onlyOwner
+    {
+        require(_clientChainGatewayLogic != address(0), "Bootstrap: client chain gateway logic address cannot be empty");
+        require(_clientChainInitializationData.length >= 4, "Bootstrap: client chain initialization data is malformed");
         clientChainGatewayLogic = _clientChainGatewayLogic;
         clientChainInitializationData = _clientChainInitializationData;
-        emit ClientChainGatewayLogicUpdated(
-            _clientChainGatewayLogic,
-            _clientChainInitializationData
-        );
+        emit ClientChainGatewayLogicUpdated(_clientChainGatewayLogic, _clientChainInitializationData);
```
If the signature of `ClientChainGateway` changes, this number would have to be updated each time. I don't think such a burden is worth the gains it provides.

MaxMustermann2 replied:
> if feasible within the constraints of Solidity and the blockchain environment

Too expensive from a gas cost perspective. Won't fix.

MaxMustermann2 commented on src/core/Bootstrap.sol:
```diff
@@ -54,19 +51,13 @@
-            spawnTime_ > offsetDuration_,
-            "Bootstrap: spawn time should be greater than offset duration"
-        );
+        require(spawnTime_ > offsetDuration_, "Bootstrap: spawn time should be greater than offset duration");
         uint256 lockTime = spawnTime_ - offsetDuration_;
+        require(lockTime > block.timestamp, "Bootstrap: lock time should be in the future");
```
Won't fix. There is nothing for a malicious client chain validator to gain here by altering the timestamp.

MaxMustermann2 commented on test/foundry/Bootstrap.t.sol:
```diff
@@ -718,19 +604,15 @@ contract BootstrapTest is Test {
-        bootstrap.delegateTo(
-            "exo13hasr43vvq8v44xpzh0l6yuym4kca98f87j7ac", address(0xa), amounts[0]
-        );
+        vm.expectRevert("BootstrapStorage: token is not whitelisted");
```
This interpretation is wrong, since `Bootstrap` inherits from `BootstrapStorage` which does have this error string.

bwhour commented on .github/workflows/fmt.yml:
```diff
@@ -0,0 +1,56 @@
+          # this is a previous recent nightly release that should work
+          version: nightly-f625d0fa7c51e65b4bf1e8f7931cd1c6e2e285e9
+
+      - name: Run Forge build
```
some efficiency optimization could save time cost:
1. we could parallelize the build and format check steps.
2. we could ache the Foundry toolchain and other dependencies to reduce download and installation time for each run.

bwhour commented on src/libraries/Endian.sol:
```diff
@@ -11,14 +11,8 @@ library Endian {
-            ((0x0000000000FF0000 & n) << 24) |
-            ((0x000000000000FF00 & n) << 40) |
-            ((0x00000000000000FF & n) << 56);
+        return (n >> 56) | ((0x00FF000000000000 & n) >> 40) | ((0x0000FF0000000000 & n) >> 24)
```
the formate of the above looks better than the formate below

MaxMustermann2 replied:
Adding this file to the ignore list.

MaxMustermann2 commented on foundry.toml:
```diff
@@ -10,5 +10,13 @@ evm_version = "paris"
 
+[fmt]
+ignore = ["src/libraries/Endian.sol"]
+# TODO: enable these after linter is merged
+# number_underscore = "thousands"
+# sort_imports = true
+# wrap_comments = true
+# single_line_statement_blocks = "multi"
+# contract_new_lines = true
+# See more config options https://github.com/foundry-rs/foundry/blob/master/crates/config/README.md#all-options
```
Yes please create it, assign it to me. Refer to PRs 16 and 17 in it. Then resolve this conversation.

adu-web3 reviewed (approved):
good jub!

MaxMustermann2 commented:
Superseded by #17
